File: packages/engine/lib/src/models/activity.dart
import 'package:meta/meta.dart';

import 'enums.dart';

/// The second currency: how much an activity covers each of the five need
/// dimensions, 0–3 (`docs/datenmodell.md`, section „Fünf Entscheidungen").
@immutable
class Needs {
  const Needs({
    required this.physical,
    required this.mentalWork,
    required this.scent,
    required this.social,
    required this.recovery,
  });

  final int physical;
  final int mentalWork;
  final int scent;
  final int social;
  final int recovery;

  int operator [](NeedDimension d) => switch (d) {
        NeedDimension.physical => physical,
        NeedDimension.mentalWork => mentalWork,
        NeedDimension.scent => scent,
        NeedDimension.social => social,
        NeedDimension.recovery => recovery,
      };
}

@immutable
class TroubleshootingEntry {
  const TroubleshootingEntry({required this.problem, required this.answer});

  final String problem;
  final String answer;
}

/// Anything that can fill a day's slot: a training session, a sniffing game,
/// a daily routine, a rest suggestion. Not every activity trains a skill.
/// Content comes from `content/aktivitaeten/*.yaml`, see
/// `content/schema/aktivitaet.yaml`.
@immutable
class Activity {
  const Activity({
    required this.id,
    required this.title,
    required this.sentence,
    required this.type,
    this.trainsSkill,
    required this.needs,
    required this.arousal,
    required this.durationMin,
    required this.durationMax,
    required this.location,
    this.forDistraction,
    required this.isRefresher,
    required this.heatSuitable,
    required this.rainSuitable,
    required this.darknessSuitable,
    required this.jointStraining,
    this.seasonalWindow,
    this.equipment = const [],
    required this.secondPerson,
    required this.minAgeWeeks,
    this.maxAgeWeeks,
    this.suitability = const {},
    required this.varianceGroup,
    required this.cooldownDays,
    this.illustration,
    this.instructions = const [],
    required this.successCriterion,
    this.commonMistakes = const [],
    this.troubleshooting = const [],
  });

  final String id;
  final String title;

  /// THE sentence for the day view, one or two lines.
  final String sentence;

  final ActivityType type;

  /// `null` for enrichment.
  final String? trainsSkill;

  final Needs needs;

  /// 0–3, how much arousal is left afterwards.
  final int arousal;

  final int durationMin;
  final int durationMax;
  final Location location;

  /// Only for `type == training`: the matching distraction levels, as
  /// [min, max].
  final (int, int)? forDistraction;

  final bool isRefresher;

  final bool heatSuitable;
  final bool rainSuitable;
  final bool darknessSuitable;
  final bool jointStraining;

  /// e.g. `[10, 11, 12]` for New Year's Eve preparation.
  final List<int>? seasonalWindow;

  final List<String> equipment;
  final bool secondPerson;
  final int minAgeWeeks;
  final int? maxAgeWeeks;

  /// Weighted, NEVER filters hard (`docs/datenmodell.md`).
  final Map<BreedGroup, int> suitability;

  /// The cooldown is tied to the variance group, not the activity — basic
  /// cues need repetition, enrichment doesn't.
  final String varianceGroup;
  final int cooldownDays;

  final String? illustration;
  final List<String> instructions;
  final String successCriterion;
  final List<String> commonMistakes;
  final List<TroubleshootingEntry> troubleshooting;
}

File: infra/supabase/functions/_shared/planner/models/activity.ts
import type { ActivityType, BreedGroup, Location, NeedDimension } from './enums.ts';

/**
 * The second currency: how much an activity covers each of the five need
 * dimensions, 0–3 (`docs/datenmodell.md`, section „Fünf Entscheidungen").
 */
export interface Needs {
  readonly physical: number;
  readonly mentalWork: number;
  readonly scent: number;
  readonly social: number;
  readonly recovery: number;
}

export function needFor(needs: Needs, dimension: NeedDimension): number {
  switch (dimension) {
    case 'physical':
      return needs.physical;
    case 'mentalWork':
      return needs.mentalWork;
    case 'scent':
      return needs.scent;
    case 'social':
      return needs.social;
    case 'recovery':
      return needs.recovery;
  }
}

export interface TroubleshootingEntry {
  readonly problem: string;
  readonly answer: string;
}

/**
 * Anything that can fill a day's slot: a training session, a sniffing game,
 * a daily routine, a rest suggestion. Not every activity trains a skill.
 * Content comes from `content/aktivitaeten/*.yaml`, see
 * `content/schema/aktivitaet.yaml`.
 */
export interface Activity {
  readonly id: string;
  readonly title: string;

  /** THE sentence for the day view, one or two lines. */
  readonly sentence: string;

  readonly type: ActivityType;

  /** `null` for enrichment. */
  readonly trainsSkill: string | null;

  readonly needs: Needs;

  /** 0–3, how much arousal is left afterwards. */
  readonly arousal: number;

  readonly durationMin: number;
  readonly durationMax: number;
  readonly location: Location;

  /** Only for `type === 'training'`: the matching distraction levels, as [min, max]. */
  readonly forDistraction: readonly [number, number] | null;

  readonly isRefresher: boolean;

  readonly heatSuitable: boolean;
  readonly rainSuitable: boolean;
  readonly darknessSuitable: boolean;
  readonly jointStraining: boolean;

  /** e.g. `[10, 11, 12]` for New Year's Eve preparation. */
  readonly seasonalWindow: readonly number[] | null;

  readonly equipment: readonly string[];
  readonly secondPerson: boolean;
  readonly minAgeWeeks: number;
  readonly maxAgeWeeks: number | null;

  /** Weighted, NEVER filters hard (`docs/datenmodell.md`). */
  readonly suitability: ReadonlyMap<BreedGroup, number>;

  /**
   * The cooldown is tied to the variance group, not the activity — basic
   * cues need repetition, enrichment doesn't.
   */
  readonly varianceGroup: string;
  readonly cooldownDays: number;

  readonly illustration: string | null;
  readonly instructions: readonly string[];
  readonly successCriterion: string;
  readonly commonMistakes: readonly string[];
  readonly troubleshooting: readonly TroubleshootingEntry[];
}
